### DISCUSSION TITLE
5 useful extensions for Task<T> in .NET

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#53] [General]
# 5 useful extensions for Task<T> in .NET

In this short blog post, I will show you 5 useful extensions for Task in .NET. We will build them as extension methods, so there are easy to use.
On top, I will show a small example of how to use them. So let's go!

https://steven-giesel.com/blogPost/d38e70b4-6f36-41ff-8011-b0b0d1f54f6e

**Comment by curia-damiano** [comment]
Very nice article, thank you!

Is there a reason why, in section 4, there are no ConfigureAwait(false) ?

**Reply by linkdotnet** [reply]
Hey - thanks @curia-damiano 
And no there is no reason - I just forgot :D.
I will add this.

**Comment by MohammadHadi2031** [comment]
Can we use [Task.WaitAsync](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitasync?view=net-7.0#system-threading-tasks-task-waitasync(system-timespan)) instead of WithTimeout in section 4?

**Reply by linkdotnet** [reply]
Yes - since .NET 6 you can use `WaitAsync`. I will amend the post! Thanks

**Comment by curia-damiano** [comment]
In section 4, I think you can also do
   await task.ConfigureAwait(false);
Btw really annoying, I know...

On Wed, Apr 12, 2023 at 7:32 AM Steven Giesel ***@***.***>
wrote:

> Yes - since .NET 6 you can use WaitAsync. I will amend the post! Thanks
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/linkdotnet/Blog.Discussions/discussions/53#discussioncomment-5588251>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEN72MYVMUAIEWSRGHJ7DHLXAY5ADANCNFSM6AAAAAAW2HHGK4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
*Damiano Curia*
Swiss mobile: +41 76 50 40 60 2
Italian mobile: +39 348/360.25.73
skype: curia.damiano
email: ***@***.***
www: http://curia.me

**Comment by rikbosch** [comment]
In case of section 2 (retry) wouldn't you want the Task continuing on the original context (if any)? Now, the first execution will run on within the original context, but the retries will be scheduled outside of the original context (somewhere in the threadpool).

In asp.net (core) you won't notice this as there is no synchronization context (https://blog.stephencleary.com/2017/03/aspnetcore-synchronization-context.html), but in a UI application or old-school asp.net this could cause issues.

**Reply by curia-damiano** [reply]
Correct, but usually it will be the caller that will re-enter the original context.
If there is a UI update inside the async call, there will be need of a Dispatcher.

i.e. the caller can't call an async method and do direct UI updates, assuming to be in the main thread - otherwise it wouldn't be an async method by definition

**Comment by molszews** [comment]
Re 4. in this implementation will the original (potentailly long running) Task be cancelled after specified timeout or not?

**Reply by linkdotnet** [reply]
This version doesn’t pass in a cancellation token. So it would still run. You could adopt it to use such a token

**Comment by slaneyrw** [comment]
section 4... be careful about using Task.Delay.  Under the hood it sets up a Timer which will only be disposed on completion of the delay... whether or not you are still awaiting the result.  If the delay is used in a loop, or relatively long lasting, you may end up using more resources than you would expect.

To my knowledge, there is no wait to dispose this internal timer if you no longer want to wait.

However there are no reasonable alternatives though.

**Comment by MohammadHadi2031** [comment]
In section 2. How about catching OperationCanceledException and prevent from retrying?
````
try
{
    ...
}
catch (OperationCanceledException)
{
    throw;
}
catch 
{
    ...
}
````

**Reply by linkdotnet** [reply]
Yes that would be a good edition! Well put

**Comment by msftgeek** [comment]
Great extensions!

But, I run into this when trying to build #4: (.NET 6)

https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs1503?f1url=%3FappId%3Droslyn%26k%3Dk(CS1503)

Severity	Code	Description	Project	File	Line	Suppression State
Error	CS1503	Argument 2: cannot convert from 'System.Runtime.CompilerServices.ConfiguredTaskAwaitable' to 'System.Threading.Tasks.Task'	ACME.CommonLibrary	C:\repos\ACME.CommonLibrary\ACME.CommonLibrary\Extensions\TaskExtensions.cs	54	Active

**Reply by linkdotnet** [reply]
Yeah that is my fault - I accidently added too many `ConfigureAwait`s ;). Fixed
